x/gateway/keeper/msg_server_stake_gateway.go

```diff
@@ -0,0 +1,91 @@
+package keeper
+
+import (
+	"context"
+	"fmt"
+
+	sdkerrors "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/pokt-network/poktroll/x/gateway/types"
+)
+
+func (k msgServer) StakeGateway(
+	goCtx context.Context,
+	msg *types.MsgStakeGateway,
+) (*types.MsgStakeGatewayResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	logger := k.Logger().With("method", "StakeGateway")
+	logger.Info(fmt.Sprintf("About to stake gateway with msg: %v", msg))
+
+	if err := msg.ValidateBasic(); err != nil {
+		return nil, err
+	}
+
+	// Check if the gateway already exists or not
+	var err error
+	var coinsToDelegate sdk.Coin
+	gateway, isGatewayFound := k.GetGateway(ctx, msg.Address)
+	if !isGatewayFound {
+		logger.Info(fmt.Sprintf("Gateway not found. Creating new gateway for address %s", msg.Address))
+		gateway = k.createGateway(ctx, msg)
+		coinsToDelegate = *msg.Stake
+	} else {
+		logger.Info(fmt.Sprintf("Gateway found. Updating gateway stake for address %s", msg.Address))
+		currGatewayStake := *gateway.Stake
+		if err = k.updateGateway(ctx, &gateway, msg); err != nil {
+			return nil, err
+		}
+		coinsToDelegate = (*msg.Stake).Sub(currGatewayStake)
```

Comment:
Am I understanding correctly that the behavior here is that the user is expected to provide the **target stake** (as opposed to the diff) and this check is to 1) only delegate the difference between an existing stake and the target and 2) ensure that the target is higher than the existing amount?

I think we may be better served by using `coin.Coin#SafeSub()`, which returns an error (instead of panicing) if the result is negative.
```suggestion
		coinsToDelegate, err = (*msg.Stake).SafeSub(currGatewayStake)
		if err != nil {
		  return nil, err
		}
```